python/lsst/ip/diffim/imageMapReduce.py

```diff
@@ -622,146 +622,121 @@ def _generateGrid(self, exposure, forceEvenSized=False, **kwargs):
         """
         # kwargs are ignored, but necessary to enable optional passing of
         # `forceEvenSized` from `_runMapper`.
+        bbox = exposure.getBBox()
 
         # Extract the config parameters for conciseness.
-        gridSizeX = self.config.gridSizeX
-        gridSizeY = self.config.gridSizeY
+        cellCentroidsX = self.config.cellCentroidsX
+        cellCentroidsY = self.config.cellCentroidsY
+        cellSizeX = self.config.cellSizeX
+        cellSizeY = self.config.cellSizeY
         gridStepX = self.config.gridStepX
         gridStepY = self.config.gridStepY
         borderSizeX = self.config.borderSizeX
         borderSizeY = self.config.borderSizeY
         adjustGridOption = self.config.adjustGridOption
         scaleByFwhm = self.config.scaleByFwhm
-        bbox = exposure.getBBox()
 
-        psfFwhm = (exposure.getPsf().computeShape().getDeterminantRadius() *
-                   2. * np.sqrt(2. * np.log(2.)))
-        if scaleByFwhm:
-            self.log.info("Scaling grid parameters by %f" % psfFwhm)
+        if cellCentroidsX is None or len(cellCentroidsX) <= 0:
+            # No given centroids; construct them from cellSize/gridStep
 
-        def rescaleValue(val):
+            psfFwhm = (exposure.getPsf().computeShape().getDeterminantRadius() *
+                       2. * np.sqrt(2. * np.log(2.)))
             if scaleByFwhm:
-                return np.rint(val * psfFwhm).astype(int)
+                self.log.info("Scaling grid parameters by %f" % psfFwhm)
+
+            def rescaleValue(val):
+                if scaleByFwhm:
+                    return np.rint(val * psfFwhm).astype(int)
+                else:
+                    return np.rint(val).astype(int)
+
+            cellSizeX = rescaleValue(cellSizeX)
+            cellSizeY = rescaleValue(cellSizeY)
+            gridStepX = rescaleValue(gridStepX)
+            gridStepY = rescaleValue(gridStepY)
+            borderSizeX = rescaleValue(borderSizeX)
+            borderSizeY = rescaleValue(borderSizeY)
+
+            nGridX = bbox.getWidth() // gridStepX
+            nGridY = bbox.getHeight() // gridStepY
+
+            if adjustGridOption == 'spacing':
+                # Readjust spacings so that they fit perfectly in the image.
+                nGridX = bbox.getWidth() // cellSizeX + 1
+                nGridY = bbox.getHeight() // cellSizeY + 1
+                cellCentroids = [(x, y) for x in np.linspace(cellSizeX//2, bbox.getWidth() - cellSizeX//2,
```

Comment:
Double-nested list comprehensions are a bit unpleasant. I think what you want to do is use `np.meshgrid` here. Possibly:

```python
np.meshgrid(np.linspace(...), np.linspace(...))
```

See: https://docs.scipy.org/doc/numpy/reference/generated/numpy.meshgrid.html